pkg/controllers/namespace/namespace_sync_controller_test.go

```diff
@@ -0,0 +1,505 @@
+	fakeRecorder := record.NewFakeRecorder(100)
+
+	c := &Controller{
+		Client:          fakeClient,
+		EventRecorder:   fakeRecorder,
+		OverrideManager: overridemanager.New(fakeClient, fakeRecorder),
+	}
+
+	err := c.buildWorks(context.Background(), namespace, clusters)
+	assert.NoError(t, err)
+
+	for _, cluster := range clusters {
+		work := &workv1alpha1.Work{}
+		err = fakeClient.Get(context.Background(), types.NamespacedName{
+			Namespace: names.GenerateExecutionSpaceName(cluster.Name),
+			Name:      names.GenerateWorkName(namespace.Kind, namespace.Name, namespace.Namespace),
+		}, work)
+		assert.NoError(t, err)
+		assert.NotNil(t, work)
+		assert.Equal(t, namespace.Name, work.OwnerReferences[0].Name)
+	}
+}
+
+func TestController_buildWorksWithOverridePolicy(t *testing.T) {
+	scheme := runtime.NewScheme()
+	_ = corev1.AddToScheme(scheme)
+	_ = clusterv1alpha1.Install(scheme)
+	_ = workv1alpha1.Install(scheme)
+	_ = policyv1alpha1.Install(scheme)
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-namespace",
+		},
+	}
+
+	clusters := []clusterv1alpha1.Cluster{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cluster1",
+			},
+		},
+	}
+
+	overridePolicy := &policyv1alpha1.ClusterOverridePolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-override",
+		},
+		Spec: policyv1alpha1.OverrideSpec{
+			ResourceSelectors: []policyv1alpha1.ResourceSelector{
+				{
+					APIVersion: "v1",
+					Kind:       "Namespace",
+					Name:       "test-namespace",
+				},
+			},
+			OverrideRules: []policyv1alpha1.RuleWithCluster{
+				{
+					TargetCluster: &policyv1alpha1.ClusterAffinity{
+						ClusterNames: []string{"cluster1"},
+					},
+					Overriders: policyv1alpha1.Overriders{
+						Plaintext: []policyv1alpha1.PlaintextOverrider{
+							{
+								Path:     "/metadata/labels/overridden",
+								Operator: "add",
+								Value: apiextensionsv1.JSON{
+									Raw: []byte(`"true"`),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(namespace, &clusters[0], overridePolicy).Build()
+	fakeRecorder := record.NewFakeRecorder(100)
+
+	c := &Controller{
+		Client:          fakeClient,
+		EventRecorder:   fakeRecorder,
+		OverrideManager: overridemanager.New(fakeClient, fakeRecorder),
+	}
+
+	err := c.buildWorks(context.Background(), namespace, clusters)
+	assert.NoError(t, err)
+
+	work := &workv1alpha1.Work{}
+	err = fakeClient.Get(context.Background(), types.NamespacedName{
+		Namespace: names.GenerateExecutionSpaceName(clusters[0].Name),
+		Name:      names.GenerateWorkName(namespace.Kind, namespace.Name, namespace.Namespace),
+	}, work)
+	assert.NoError(t, err)
+	assert.NotNil(t, work)
+
+	t.Logf("Work found: %+v", work)
+	t.Logf("Work annotations: %v", work.Annotations)
+	t.Logf("Work spec: %+v", work.Spec)
+
+	assert.NotNil(t, work)
+}
+
+func TestController_SetupWithManager(t *testing.T) {
```

Comment:
Can the following three functions, including this one, be organized into table-driven tests? That would be easier to read and maintain. I do have tests in the current style of testing, but they don't seem to work very well at the moment.

Reply:
Yeah, that would be more readable and maintainable.